Mathlib/Combinatorics/Enumerative/Polya.lean

```diff
@@ -0,0 +1,112 @@
+/-
+Copyright (c) 2025 Zihui Bai. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Zihui Bai, Zhengfeng Yang
+-/
+
+module
+
+public import Mathlib.Data.Fintype.Basic
+public import Mathlib.GroupTheory.GroupAction.Basic
+public import Mathlib.GroupTheory.GroupAction.Quotient
+
+/-!
+## Main definitions and results
+
+* `MulAction (Equiv.Perm X) (X → Y)`: action of permutations on colorings by precomposition.
+* `coloringEquiv`: two colorings are equivalent if they lie in the same orbit.
+* `coloringEquiv_equivalence`: this defines an equivalence relation.
+* `orbit_size_eq_index`: the orbit–stabilizer theorem reformulated for colorings.
+
+## Implementation notes
+
+We model a coloring as a function `X → Y`, where `X` is a finite set of objects
+and `Y` is a finite set of colors.
+The permutation group `Equiv.Perm X` acts on colorings by precomposition.
+-/
+
+@[expose] public section
+
+open scoped BigOperators
+open MulAction Finset Equiv
+
+namespace DistinctColorings
+
+variable {X Y : Type*} (c : X → Y)
+
+/--
+A *coloring* is a function `X → Y`.
+The permutation group `Equiv.Perm X` acts on colorings by precomposition:
+`(g • c) x = c (g⁻¹ x)`.
+-/
+instance : MulAction (Equiv.Perm X) (X → Y) where
+  smul g c := fun x ↦ c (g⁻¹ • x)
+  one_smul := by
+    intro c
+    ext x
+    simp [HSMul.hSMul]
+    change c ((1 : Equiv.Perm X) • x) = c x
+    rw [one_smul (Equiv.Perm X) x]
+  mul_smul := by
+    intro g g' f
+    ext x
+    simp [HSMul.hSMul]
+    change f ((g'⁻¹ * g⁻¹) • x) = f (g'⁻¹ • (g⁻¹ • x))
+    rw [mul_smul g'⁻¹ g⁻¹ x]
+
+/--
+For permutations `f, g : Equiv.Perm X`, the colorings `g • c` and `f • c`
+are equal if and only if `f⁻¹ * g` belongs to the stabilizer of `c`.
+-/
+theorem smul_eq_iff_mem_stabilizer (f g : Equiv.Perm X) (c : X → Y) :
+    g • c = f • c ↔ f⁻¹ * g ∈ MulAction.stabilizer (Equiv.Perm X) c := by
```

Comment:
This lemma is not general enough for mathlib: it could be generalized to the following
```
lemma smul_eq_iff_mem_stabilizer {G α} [Group G] [MulAction G α] {g h : G} {x : α} :
    g • x = h • x ↔ h⁻¹ * g ∈ MulAction.stabilizer G x := by
  rw [mem_stabilizer_iff, mul_smul, inv_smul_eq_iff, eq_comm]

lemma DomMulAct.mk_smul_eq_iff_mem_stabilizer {G α β} [Group G] [MulAction G α]
    {g h : G} {x : α → β} :
    DomMulAct.mk g • x = DomMulAct.mk h • x ↔
    DomMulAct.mk (g * h⁻¹) ∈ MulAction.stabilizer Gᵈᵐᵃ x :=
  smul_eq_iff_mem_stabilizer
```
which are not currently in mathlib as far as I can see.